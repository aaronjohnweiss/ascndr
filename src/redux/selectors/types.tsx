import { Data, OrderedList, Persisted } from '../../types/Firebase'
import { Gym } from '../../types/Gym'
import { Route } from '../../types/Route'
import { Session } from '../../types/Session'
import { User } from '../../types/User'
import { Workout } from '../../types/Workout'
import { DatabaseState } from './selectors'
import { Goal } from '../../types/Goal'

export type Optional<T> = T | undefined
/**
 * Predicate that can evaluate to true or false, but also return undefined to indicate that data is still loading
 */
export type FilterPredicate<T> = (t: T) => Optional<boolean>
/**
 * Function that takes filter arguments and returns a FilterPredicate
 */
type FilterFunction<T> = (values: Optional<string[]>) => FilterPredicate<T>
/**
 * Object that can be filtered upon; should define a constant _type for disambiguation
 */
export type Filterable = Gym | Route | Session | User | Workout | Goal

/**
 * All filters that can be provided for each type
 */
const FilterableFields = {
  gym: ['owner', 'editor', 'viewer', 'gymKey'],
  route: ['gym', 'session', 'viewer'],
  session: ['gym', 'owner', 'viewer', 'route', 'sessionKey'],
  user: ['uid', 'friendOf'],
  workout: ['owner', 'viewer'],
  goal: ['owner', 'viewer'],
} as const
/**
 * All applicable filters for a given Filterable type
 */
type FilterName<T extends Filterable> = (typeof FilterableFields)[T['_type']][number]
/**
 * A Filter defines FilterFunctions for every FilterName for a given Filterable type
 * (essentially, an implementation of all possible filters on a given type)
 */
export type Filter<T extends Filterable> = Record<FilterName<T>, FilterFunction<Persisted<T>>>
/**
 * A StateFilter wraps a filter in a closure that has access to the database state
 */
export type StateFilter<T extends Filterable> = (databaseState: DatabaseState) => Filter<T>
/**
 * Caller-supplied filter value
 */
export type FilterValue = Optional<string | string[]>
/**
 * Caller-supplied filter parameter (filter name and value)
 */
export type FilterParam<T extends Filterable> = [FilterName<T>, FilterValue]
/**
 * Function that will take in filter parameters and return the relevant data
 */
export type ParameterizedSelector<
  Whole extends Filterable,
  WrappedData extends OrderedList<Whole> | Data<Whole>,
> = (...params: FilterParam<Whole>[]) => Optional<WrappedData>

/**
 * Selectors to expose to the caller
 */
export interface Selectors<T extends Filterable> {
  /**
   * Return an array containing the requested data
   */
  getOrdered: ParameterizedSelector<T, OrderedList<T>>
  /**
   * Return a map from key->value for the requested data
   */
  getData: ParameterizedSelector<T, Data<T>>
  /**
   * Find the entry with the given id, if it exists
   */
  getOne: (id: string) => Optional<T>
}
