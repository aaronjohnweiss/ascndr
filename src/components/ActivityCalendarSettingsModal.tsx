import React, { useState } from 'react'
import { Button, Form, Modal } from 'react-bootstrap'
import { getUserName } from '../helpers/filterUtils'
import { MultiSelect } from '../containers/StatFilters'
import {
  ActivityCalendarPreferences,
  CalendarMode,
  defaultActivityCalendarPreferences,
  User,
} from '../types/User'

export const getPreferences = (user: User): ActivityCalendarPreferences => ({
  ...defaultActivityCalendarPreferences(),
  ...((user && user.preferences && user.preferences.activityCalendar) || {}),
})

export const ActivityCalendarSettingsModal = ({
  user,
  friends,
  show,
  handleClose,
  handleSubmit,
}) => {
  const [preferences, setPreferences] = useState(getPreferences(user))

  const setPreference = name => val => setPreferences({ ...preferences, [name]: val })

  const userOptions = friends.map(u => ({
    key: u.uid,
    label: getUserName(u),
    checked: preferences.friends.includes(u.uid),
  }))

  return (
    <Modal show={show} onHide={handleClose}>
      <Form>
        <Modal.Header closeButton>
          <Modal.Title>Activity Calendar Settings</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form.Check
            checked={preferences.mode === CalendarMode.FRIENDS}
            label={'Include friends'}
            onChange={evt =>
              setPreference('mode')(
                evt.target.checked ? CalendarMode.FRIENDS : CalendarMode.USER_ONLY
              )
            }
          />

          <Form.Check
            checked={preferences.includeWorkouts}
            label={'Include workouts'}
            onChange={evt => setPreference('includeWorkouts')(evt.target.checked)}
          />

          {preferences.mode === CalendarMode.USER_ONLY && preferences.includeWorkouts && (
            <Form.Check
              checked={preferences.splitWorkouts}
              label={'Split workouts by type'}
              onChange={evt => setPreference('splitWorkouts')(evt.target.checked)}
            />
          )}

          {preferences.mode === CalendarMode.FRIENDS && (
            <>
              <Form.Label>Friends to Include</Form.Label>
              <MultiSelect options={userOptions} onChange={setPreference('friends')} />
            </>
          )}
        </Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={handleClose}>
            Cancel
          </Button>
          <Button variant="primary" onClick={() => handleSubmit(preferences)}>
            Submit
          </Button>
        </Modal.Footer>
      </Form>
    </Modal>
  )
}
