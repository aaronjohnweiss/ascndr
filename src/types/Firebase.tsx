export type Persisted<T> = {key: string, value: T}
export type OrderedList<T> = Persisted<T>[]
export type Data<T> = Record<string, T>