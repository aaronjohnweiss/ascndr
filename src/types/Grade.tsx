import { ALL_MODIFIERS, ALL_STYLES } from '../helpers/gradeUtils'

export type RouteStyle = (typeof ALL_STYLES)[number]
export type RouteModifier = (typeof ALL_MODIFIERS)[number]

export const isStyle = (str: string): str is RouteStyle =>
  ALL_STYLES.find(style => style === str) !== undefined
export interface Grade {
  difficulty: number
  style: RouteStyle
  modifier?: RouteModifier
}

export interface GradeDecoration {
  percentage?: number
}

export type DecoratedGrade = Grade & GradeDecoration

export type DecoratedCustomGrade = GradeDecoration & { key: string }
