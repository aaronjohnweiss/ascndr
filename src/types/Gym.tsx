export interface Gym {
    name: string
    owner: string
    BOULDER_HEIGHT: string
    LEAD_HEIGHT: string
    TOP_ROPE_HEIGHT: string
    location?: string
}