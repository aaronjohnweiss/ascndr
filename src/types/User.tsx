export interface Hiatus {
    endDate: number
    startDate: number
}

export const CalendarMode = {
    USER_ONLY: 'User only',
    FRIENDS: 'Include friends',
} as const

export interface ActivityCalendarPreferences {
    friends: string[]
    includeWorkouts: boolean
    mode: typeof CalendarMode[keyof typeof CalendarMode]
    splitWorkouts: boolean
}

export interface UserPreferences {
    activityCalendar: ActivityCalendarPreferences
}

export interface User {
    name?: string
    friends: string[]
    hiatuses: Hiatus[]
    preferences: UserPreferences
    uid: string
}

export type FirebaseUser = Pick<User, 'uid'> & Partial<User>

export const defaultUser = (part: FirebaseUser): User => ({
    friends: [],
    hiatuses: [],
    preferences: {
        activityCalendar: defaultActivityCalendarPreferences()
    },
    name: part.uid,
    ...part
})
export const missingUser = (uid: string): User => defaultUser({uid})

export const defaultActivityCalendarPreferences = (): ActivityCalendarPreferences => ({
    mode: CalendarMode.USER_ONLY,
    includeWorkouts: true,
    splitWorkouts: true,
    friends: [],
})