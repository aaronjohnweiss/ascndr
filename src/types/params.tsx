export type IdParams = {
  id: string
}
