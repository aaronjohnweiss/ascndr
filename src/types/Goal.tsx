import { WorkoutCategory } from './Workout'
import { Grade, RouteStyle } from './Grade'

export const GOAL_CATEGORIES = [
  'ACTIVITY_COUNT',
  'WORKOUT_COUNT',
  'SESSION_COUNT',
  'ROUTE_COUNT',
  'ROUTE_DISTANCE',
] as const

export type GoalCategory = (typeof GOAL_CATEGORIES)[number]

export const isGoalCategory = (str: string): str is GoalCategory =>
  GOAL_CATEGORIES.some(cat => cat === str)

export type ActivityGoalDetails = {
  category: 'ACTIVITY_COUNT'
}

export type WorkoutGoalDetails = {
  category: 'WORKOUT_COUNT'
  workoutCategories: WorkoutCategory[]
  minIntensity: number
}

export type SessionGoalDetails = {
  category: 'SESSION_COUNT'
  minDurationMinutes: number
  minRouteCount: number
}

export type RouteCountGoalDetails = {
  category: 'ROUTE_COUNT'
  styles: Partial<
    Record<
      RouteStyle,
      {
        minGrade: Grade
      }
    >
  >
}

export type DistanceGoalDetails = {
  category: 'ROUTE_DISTANCE'
  styles: Record<
    RouteStyle,
    {
      minGrade: Grade
    }
  >
}

export type SharedGoal = {
  isShared: true
  participants: string[]
}

export type SoloGoal = {
  isShared: false
}

export type GoalDetails =
  | ActivityGoalDetails
  | WorkoutGoalDetails
  | SessionGoalDetails
  | RouteCountGoalDetails
  | DistanceGoalDetails

export type Goal = {
  _type: 'goal'
  category: GoalCategory
  owner: string
  startTime: number
  endTime: number
  target: number
  participants: string[]
} & (SharedGoal | SoloGoal) &
  GoalDetails

export type FirebaseGoal = Goal

export const isShared = (goal: Goal): goal is Goal & SharedGoal => goal.isShared
export const isWorkout = (goal: Goal): goal is Goal & WorkoutGoalDetails =>
  goal.category === 'WORKOUT_COUNT'
