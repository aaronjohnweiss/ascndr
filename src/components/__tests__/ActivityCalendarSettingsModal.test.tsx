import { CalendarMode, defaultUser } from '../../types/User'
import { ActivityCalendarSettingsModal } from '../ActivityCalendarSettingsModal'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

describe('activity calendar settings', () => {
  const friends = Array.from({ length: 5 }, (_, i) =>
    defaultUser({ uid: `uid${i}`, name: `user ${i}` }),
  )

  const user = defaultUser({
    uid: 'user',
    friends: friends.map(f => f.uid),
  })

  const handleClose = jest.fn()
  const handleSubmit = jest.fn()
  const defaultProps = () => ({
    user,
    friends,
    show: true,
    handleClose,
    handleSubmit,
  })

  it('should render with default preferences', () => {
    const wrapper = shallow(<ActivityCalendarSettingsModal {...defaultProps()} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should render with custom user preferences', () => {
    const props = {
      ...defaultProps(),
      user: {
        ...user,
        preferences: {
          activityCalendar: {
            friends: user.friends.slice(0, 2),
            includeWorkouts: true,
            mode: CalendarMode.FRIENDS,
            splitWorkouts: false,
          },
        },
      },
    }

    const wrapper = shallow(<ActivityCalendarSettingsModal {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should be hidden when passed show=false', () => {
    const props = {
      ...defaultProps(),
      show: false,
    }

    const wrapper = shallow(<ActivityCalendarSettingsModal {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should update user state based off of form input', () => {
    const props = {
      ...defaultProps(),
      user: {
        ...user,
        preferences: {
          activityCalendar: {
            mode: CalendarMode.USER_ONLY,
            includeWorkouts: false,
            friends: [],
          },
        },
      },
    }
    const wrapper = shallow(<ActivityCalendarSettingsModal {...props} />)

    wrapper
      .find('FormCheck[label="Include friends"]')
      .simulate('change', { target: { checked: true } })
    wrapper
      .find('FormCheck[label="Include workouts"]')
      .simulate('change', { target: { checked: true } })

    const selectedUids = user.friends.slice(0, 2)
    wrapper.find('MultiSelect').simulate('change', selectedUids)

    wrapper.find('Button[variant="primary"]').simulate('click')

    expect(handleSubmit).toHaveBeenCalledWith(
      expect.objectContaining({
        mode: CalendarMode.FRIENDS,
        includeWorkouts: true,
        splitWorkouts: true,
        friends: selectedUids,
      }),
    )
  })
})
