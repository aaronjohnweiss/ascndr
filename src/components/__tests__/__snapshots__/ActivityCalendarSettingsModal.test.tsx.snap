// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activity calendar settings should be hidden when passed show=false 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "ModalDialog",
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={true}
  onHide={[MockFunction]}
  restoreFocus={true}
  show={false}
>
  <Form>
    <ModalHeader
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle>
        Activity Calendar Settings
      </ModalTitle>
    </ModalHeader>
    <ModalBody>
      <FormCheck
        checked={false}
        label="Include friends"
        onChange={[Function]}
      />
      <FormCheck
        checked={true}
        label="Include workouts"
        onChange={[Function]}
      />
      <FormCheck
        checked={true}
        label="Split workouts by type"
        onChange={[Function]}
      />
    </ModalBody>
    <ModalFooter>
      <Button
        active={false}
        disabled={false}
        onClick={[MockFunction]}
        variant="secondary"
      >
        Cancel
      </Button>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Submit
      </Button>
    </ModalFooter>
  </Form>
</Modal>
`;

exports[`activity calendar settings should render with custom user preferences 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "ModalDialog",
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={true}
  onHide={[MockFunction]}
  restoreFocus={true}
  show={true}
>
  <Form>
    <ModalHeader
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle>
        Activity Calendar Settings
      </ModalTitle>
    </ModalHeader>
    <ModalBody>
      <FormCheck
        checked={true}
        label="Include friends"
        onChange={[Function]}
      />
      <FormCheck
        checked={true}
        label="Include workouts"
        onChange={[Function]}
      />
      <FormLabel
        column={false}
        visuallyHidden={false}
      >
        Friends to Include
      </FormLabel>
      <MultiSelect
        onChange={[Function]}
        options={
          Array [
            Object {
              "checked": true,
              "key": "uid0",
              "label": "user 0",
            },
            Object {
              "checked": true,
              "key": "uid1",
              "label": "user 1",
            },
            Object {
              "checked": false,
              "key": "uid2",
              "label": "user 2",
            },
            Object {
              "checked": false,
              "key": "uid3",
              "label": "user 3",
            },
            Object {
              "checked": false,
              "key": "uid4",
              "label": "user 4",
            },
          ]
        }
      />
    </ModalBody>
    <ModalFooter>
      <Button
        active={false}
        disabled={false}
        onClick={[MockFunction]}
        variant="secondary"
      >
        Cancel
      </Button>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Submit
      </Button>
    </ModalFooter>
  </Form>
</Modal>
`;

exports[`activity calendar settings should render with default preferences 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  dialogAs={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "displayName": "ModalDialog",
      "render": [Function],
    }
  }
  enforceFocus={true}
  keyboard={true}
  onHide={[MockFunction]}
  restoreFocus={true}
  show={true}
>
  <Form>
    <ModalHeader
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle>
        Activity Calendar Settings
      </ModalTitle>
    </ModalHeader>
    <ModalBody>
      <FormCheck
        checked={false}
        label="Include friends"
        onChange={[Function]}
      />
      <FormCheck
        checked={true}
        label="Include workouts"
        onChange={[Function]}
      />
      <FormCheck
        checked={true}
        label="Split workouts by type"
        onChange={[Function]}
      />
    </ModalBody>
    <ModalFooter>
      <Button
        active={false}
        disabled={false}
        onClick={[MockFunction]}
        variant="secondary"
      >
        Cancel
      </Button>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Submit
      </Button>
    </ModalFooter>
  </Form>
</Modal>
`;
